### ISSUE TITLE
query language

### ISSUE THREAD

**Issue by gedw99** [hidal-go/hidalgo#2]
Is there any plan to incorporate a query language ?

**Comment by dennwc** [comment]
Yes, the main intention is not only to provide read-write abstraction but also allow to build complex queries that backends can optimize to any native query language that is supported.

I will leave this issue open until it's clarified in the README.

**Comment by gedw99** [comment]
Could we discuss ?

Is the query language strong or weak typed ?
Is a graphql with dynamic resolvers via a AST parser on the cards ?
There are already many good graphql parsers using AST.

**Comment by dennwc** [comment]
The plan is to support at least SQL and GraphQL.

At least on the tuplestore level, the query language can be strongly typed (schema is known). 

And yes, I already implemented a dynamic GraphQL for Cayley that uses an AST to build a query plan instead of calling resolvers.

**Comment by gedw99** [comment]
thanks @dennwc 
Sounds great ! 
Is there any code in cayley to use hidalgo yet ? Would like to explore what i can do with hidalgo and graphql.

**Comment by dennwc** [comment]
The code in questions can be found here: https://github.com/cayleygraph/cayley/tree/master/query/graphql, but it's only useful for Cayley. May serve as an example how to write the similar query planner, though.

There is also a branch in Cayley that works with Hidanlgo, but it only uses the KV layer, not the tuple-store.

In any case, I'm leaving this open to serve as a starting point for further discussions.
